### PR TITLE
SummaryReplayStrategyMixin

### PR DESCRIPTION
**Pull request by dianzrong**
This is #56 but updated with instructions (changes will be italicized).

Summary (#24):
This pull request summarizes the OCR and ASCII text of the current and previous screenshots. Then, the summaries are compared, to see if the screen has changed. The change is then noted in the prompt.

Instructions:
```
git remote add dianzrong https://github.com/dianzrong/puterbot.git
git fetch dianzrong
git checkout expand-on-demo-replay-strategy
pip install -r requirements.txt --no-cache-dir --use-pep517 # new libraries were added to the requirements
python -m puterbot.record "testing out DemoReplayStrategy" # to create the recording to replay
python -m puterbot.replay DemoReplayStrategy
```

Assumptions:
- The screen can summarized by _well by 1 sentence_. This may not be true for more complex screenshots such as those with multiple windows open simultaneously.
- 50% similarity is a good indication of whether 2 screenshots are similar or not

External Libraries:

- _Sumy_
- NLTK
- _fuzzywuzzy_

Next Steps:
- testing
- _improve OCR summarizations by removing irrelevant words_

### CODE REVIEW

**Comment by abrichr** [comment]
Thank you @dianzrong ! Should we close https://github.com/MLDSAI/puterbot/pull/56 ?

From that thread:

> Excellent work! Can you please refactor this into a SummaryReplayStrategyMixin, e.g. in strategies/summary with a get_summary method?
> What is next here?

**Comment by abrichr** [comment]
Please also remove the noop alembic script.

**Comment by abrichr** [comment]
Can you please remove the unnecessary formatting changes? e.g. see https://github.com/psf/black

Please also add a test.

File: puterbot/strategies/summary_mixin.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        ascii_similarity = compare_text(cleaned_ascii_curr, cleaned_ascii_prev)
+        ocr_similarity = compare_text(ocr_curr, ocr_prev)
+
+        similarity_list = [ascii_similarity, ocr_similarity]
+
+        # may want the required minimum similarity
+        return statistics.fmean(similarity_list)
+
+
+def clean_ascii(
```
Comment (abrichr) [inline puterbot/strategies/summary_mixin.py:71]: Why is cleaning necessary for summarization?

**Comment by abrichr** [comment]
@dianzrong any update on this?

File: puterbot/strategies/summary_mixin.py (hunk 1 above)
Comment (dianzrong) [reply puterbot/strategies/summary_mixin.py:71]: @abrichr Clean ascii is to remove any strings that aren't words so the ascii text can be properly summarized

**Comment by dianzrong** [comment]
@abrichr This is ready for review

File: puterbot/strategies/summary_mixin.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+"""
+Implements a ReplayStrategy mixin which identifies whether windows are different
```
Comment (abrichr) [inline puterbot/strategies/summary_mixin.py:2]: @dianzrong while summarization may be useful, I don't think we will be using it for identifying whether windows are different for the reasons we previously discussed. We may still use it for general purpose summarization, however.

Can you please update this documentation?

File: puterbot/strategies/summary_mixin.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        super().__init__(recording)
+
+    def get_summary(
+        self,
+        ascii_curr: str,
+        ascii_prev: str,
+        ocr_curr: str,
+        ocr_prev: str,
+    ) -> float:
+        """Returns how similar the contents of 2 screenshots are.
```
Comment (abrichr) [inline puterbot/strategies/summary_mixin.py:45]: Hi @dianzrong , it seems that the method name does not match the behavior of the function or the docstring. `get_summary` is something we want, but measuring how similar the contents are based on the summary of the text contained in them is not.

Can you please modify this to simply return the summary of some text that is passed in? No need for ASCII or OCR anywhere in this file.

**Comment by dianzrong** [comment]
@abrichr Made the requested changes!

**Comment by abrichr** [comment]
@dianzrong can you please fix merge conflicts so that we can merge this?

**Comment by dianzrong** [comment]
@abrichr Resolved merge conflicts!

File: tests/openadapt/test_summary_mixin.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+REPLAY = DemoReplayStrategy(RECORDING)
+
+def test_summary_empty():
+    empty_text = ""
+    actual = REPLAY.get_summary(empty_text, 1)
+    assert len(actual) == 0
+
+
+def test_summary_sentence():
+    story = "However, this bottle was not marked “poison,” so Alice ventured to taste it, \
```
Comment (abrichr) [inline tests/openadapt/test_summary_mixin.py:19]: 😄

**Pull request merged**